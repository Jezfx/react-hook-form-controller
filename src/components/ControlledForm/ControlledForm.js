import React, { useMemo } from "react";
import { FormProvider } from "react-hook-form";
import { Grid } from "@material-ui/core";

import FollowUpField from "./FollowUpField";
import {
  hasFollowUp,
  getName,
  getFollowUp,
  formatValues,
} from "./ControlledForm.utils";

export const renderFields = (fields) =>
  fields.map(({ Component, ...props }, i) => (
    <Grid key={getName(props) || i} item>
      <Component {...props} />

      {hasFollowUp(props) && (
        <FollowUpField
          key={`${getName(props)}-followUp`}
          fieldName={getName(props)}
          followUp={getFollowUp(props)}
        />
      )}
    </Grid>
  ));

const ControlledForm = ({ onSubmit, buttons, methods, fields }) => {
  const { handleSubmit } = methods;
  const renderFormFields = useMemo(() => renderFields(fields), [fields]);
  const renderButtons = useMemo(() => (!!buttons ? buttons() : null), [
    buttons,
  ]);

  return (
    <FormProvider {...methods}>
      <form onSubmit={handleSubmit((values) => onSubmit(formatValues(values)))}>
        <Grid spacing={3} direction="column" container>
          {renderFormFields}
          {renderButtons}
        </Grid>
      </form>
    </FormProvider>
  );
};

ControlledForm.defaultProps = {
  fields: [],
};

export default ControlledForm;
