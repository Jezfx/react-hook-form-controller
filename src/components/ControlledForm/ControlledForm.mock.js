import ControlledTextField from "../ControlledTextField";
import ControlledRadioGroup from "../ControlledRadioGroup";

export const defaultFields = [
  {
    Component: ControlledTextField,
    name: "firstName",
    label: "Text Field",
  },
];

export const followUpFields = [
  {
    Component: ControlledRadioGroup,
    name: "level_one",
    label: "Level one (check yes to see level 2)",
    options: [
      {
        name: "isSameName",
        label: "Yes",
        value: "yes",
      },
      {
        name: "isDifferentName",
        label: "No",
        value: "no",
      },
    ],
    followUp: {
      condition: () => true,
      fields: [
        {
          label: "Other Options...",
          Component: ControlledRadioGroup,
          name: "other_options",
          options: [
            {
              name: "isSameName",
              label: "Follow up",
              value: "yes",
            },
          ],
        },
      ],
    },
  },
];
