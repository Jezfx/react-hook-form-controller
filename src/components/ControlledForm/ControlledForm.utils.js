import { isNotEmpty } from "../../utils";
import { get, flow, without, isArray } from "lodash/fp";
import { transform } from "lodash";

export const getName = get("name");
export const getFollowUp = get("followUp");
export const hasFollowUp = flow([getFollowUp, isNotEmpty]);

const omitUndefinedAndNull = without([undefined, null]);
const formatArray = flow(omitUndefinedAndNull);

export const formatValues = (values) =>
  transform(
    values,
    function (result, value, key) {
      const formatedValues = isArray(value) ? formatArray(value) : value;

      return (result[key] = formatedValues);
    },
    {}
  );
