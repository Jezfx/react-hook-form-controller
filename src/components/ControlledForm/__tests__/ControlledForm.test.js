import React from "react";
import {
  waitFor,
  fireEvent,
  cleanup,
  render,
  screen,
} from "@testing-library/react";
import { Button } from "@material-ui/core";
import { useForm } from "react-hook-form";

import ControlledForm from "../";
import { defaultFields, followUpFields } from "../ControlledForm.mock";

const withFormWrapper = (Component) => (componentProps, defaultValues = {}) => {
  const methods = useForm({ defaultValues });
  return <Component methods={methods} {...componentProps} />;
};

describe("Controlled Form", () => {
  afterEach(cleanup);

  const props = {
    buttons: () => <Button type="submit">Submit</Button>,
    onSubmit: jest.fn(),
  };

  const CompWithForm = withFormWrapper(ControlledForm);
  it("should render correctly", () => {
    const { asFragment } = render(
      <CompWithForm fields={defaultFields} {...props} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should call onsubmit when submit button is clicked", async () => {
    render(<CompWithForm fields={defaultFields} {...props} />);

    const button = screen.getByText("Submit");
    fireEvent.click(button);

    await waitFor(() => {
      expect(props.onSubmit).toHaveBeenCalled();
    });
  });

  it("should render follow up fields when condition is true", () => {
    render(<CompWithForm fields={followUpFields} {...props} />);
    const followUp = screen.getByLabelText("Follow up");

    expect(followUp).toBeInTheDocument();
  });
});
