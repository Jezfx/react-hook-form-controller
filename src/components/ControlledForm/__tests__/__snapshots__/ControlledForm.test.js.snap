// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controlled Form should render correctly 1`] = `
<DocumentFragment>
  <form>
    <div>
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
          data-shrink="false"
        >
          Text Field
        </label>
        <div
          class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiFilledInput-input"
            name="firstName"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiButton-label"
      >
        Submit
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
